### PR TITLE
Add Slice Function

### PR DESCRIPTION
**Pull request by svozza**
As discussed in #39. This actually turned out to be more difficult than anticipated, I think I've managed to test all the edge cases.

### CODE REVIEW

**Comment by davidchambers** [comment]
Thanks! I'm away from home this weekend, and don't like to review code on my phone. I'll have a proper look on Monday. :)

**Comment by svozza** [comment]
Actually, probably for the best you're not looking at it now, just realised I completely forgot to add documentation!

**Comment by svozza** [comment]
Just thinking, do we want to do a check with `R.nil` as well, give that this sort of thing gives a result:

```javascript
R.slice(null, 2, [1, 2, 3, 4, 5])
// => [1, 2]

R.slice(void 0, 2, [1, 2, 3, 4, 5])
// => [1, 2]
```

**Comment by davidchambers** [comment]
I'd prefer to leave the behaviour of non-number index values unspecified.

File: index.js [hunk 1]
```diff
@@ -942,7 +977,7 @@
   //. > S.init([])
   //. Nothing()
   //. ```
-  S.init = R.ifElse(R.isEmpty, Nothing, R.compose(Just, R.init));
+  S.init = S.slice(0, -1);
```
Comment (davidchambers) [inline index.js]: Let's accept `-0` to indicate the position half a step beyond the last index. Currently we have:

```javascript
 [ 'a' , 'b' , 'c' , 'd' , 'e' ]

 0     1     2     3     4     5
-5    -4    -3    -2    -1
```

Instead we could have:

```javascript
 [ 'a' , 'b' , 'c' , 'd' , 'e' ]

 0     1     2     3     4     5
-5    -4    -3    -2    -1    -0
```

This would facilitate this symmetry:

```javascript
S.tail = S.slice(1, -0);
S.init = S.slice(0, -1);
```

With `R.slice` one can use `Infinity` as the end index, but that does not work when the index needs to be within the array's bounds.

The most straightforward expression to detect `-0` is `R.toString(x) === '-0'`.

File: index.js [hunk 2]
```diff
@@ -886,6 +886,41 @@
     return idx >= 0 && idx < len ? Just(xs[idx]) : Nothing();
   });
 
+  //# slice :: Number -> Number -> [a] -> Maybe [a]
+  //.
+  //. Returns Just a list containing the elements from the supplied list
+  //. from a beginning index (inclusive) to an end index (exclusive).
+  //. Returns Nothing if the range specified by the two indices exceeds
+  //. length of the list. Accepts negative indices, which indicate an
```
Comment (davidchambers) [inline index.js]: It's not accurate to write _exceeds length of the list_. Perhaps we could write _Returns Nothing unless the start interval is less than or equal to the end interval, and the list contains both (half-open) intervals._

File: index.js [hunk 3]
```diff
@@ -886,6 +886,41 @@
...
+  //.
+  //. > S.slice(2, -1, [1, 2, 3, 4, 5])
+  //. Just([3, 4])
+  //.
+  //. > S.slice(0, 2, [])
+  //. Nothing()
+  //.
+  //. > S.slice(1, 6, [1, 2, 3, 4, 5])
+  //. Nothing()
+  //. ```
```
Comment (davidchambers) [inline index.js]: Let's use `['a', 'b', 'c', 'd', 'e']` as the example list.

File: index.js (hunk 1 above)
Comment (CrossEye) [inline index.js]: That's really elegant!

I've not really looked, but I haven't seen many truly useful reasons to have `-0` except that the number implementation demanded it.  This shows that there really can be such.  Nice!

Comment (svozza) [inline index.js]: This also allows us to define `S.at` in terms on `S.slice` too.

Comment (davidchambers) [inline index.js]: > This also allows us to define `S.at` in terms on `S.slice` too.

I don't see it, but I'll wait to see your implementation. :)

The important thing is that `S.at(-0, xs)` should return a Nothing for every `xs`, since in this context `-0` refers to the position *after* the last index.

Comment (svozza) [inline index.js]: The implementation is already in there. :) (Although I had forgotten to add a test for the -0 functionality in the `S.at` tests.)

Comment (svozza) [inline index.js]: Also, I started a points free implementation but gave up  as soon as I came up with this abomination:

```javascript
var at = S.at = R.curry(function(n, xs) {
  R.compose(R.map(R.head), S.slice(n, R.__, xs), R.ifElse(R.compose(R.eq(0), R.inc), R.always(-0), R.inc))(n);
});
```

:smile:

File: index.js [hunk 4]
```diff
@@ -881,9 +881,47 @@
   //. Just("d")
   //. ```
   var at = S.at = R.curry(function(n, xs) {
+    var endIndex = R.eq(0, n + 1) ? -0 : n + 1;
```
Comment (davidchambers) [inline index.js]: Could `R.eq(0, n + 1)` be written as `n === -1`?

File: index.js (hunk 1 above)
Comment (davidchambers) [inline index.js]: Yikes!

File: index.js (hunk 4 above)
Comment (svozza) [inline index.js]: It could indeed, I was trying to be too clever there.

File: index.js (hunk 1 above)
Comment (svozza) [inline index.js]: I guess it's not as bad in the `R.compose(R.eq(0), R.inc)` is replaced with R.eq(-1).

File: index.js [hunk 5]
```diff
@@ -881,9 +881,47 @@
   //. Just("d")
   //. ```
   var at = S.at = R.curry(function(n, xs) {
+    var endIndex = n === -1 ? -0 : n + 1;
+    return R.compose(R.map(R.head), S.slice(n, endIndex))(xs);
```
Comment (davidchambers) [inline index.js]: When we use a Sanctuary function internally it's nice to use a local alias (`slice` in this case). Also, I don't see the value of `R.compose` here: since we're applying the composition immediately, we might as well nest function calls.

```javascript
return R.map(R.head, slice(n, n === -1 ? -0 : n + 1, xs));
```

File: test/index.js [hunk 6]
```diff
@@ -1106,6 +1106,10 @@ describe('list', function() {
       assert(S.at(-1, ['foo', 'bar', 'baz']).equals(S.Just('baz')));
     });
 
+    it('returns Nothing when passed the index -0', function() {
+      assert(S.at(-0, ['foo', 'bar', 'baz']).equals(S.Nothing()));
+    });
```
Comment (davidchambers) [inline test/index.js]: I suggest we move this assertion to the following block, since `-0` is an out-of-bounds index.

File: index.js (hunk 5 above)
Comment (svozza) [inline index.js]: Yeah, you're right the `compose` is completely unnecessary. Force of habit!

File: test/index.js (hunk 6 above)
Comment (svozza) [inline test/index.js]: Makes sense.

File: index.js [hunk 7]
```diff
@@ -864,6 +864,45 @@
...
+  //. ```
+  var slice = S.slice = R.curry(function(start, end, xs) {
+    var len = xs.length;
+    var startIdx = R.eq(-0, start) ? len : start < 0 ? start + len : start;
+    var endIdx = R.eq(-0, end) ? len : end < 0 ? end + len : end;
+
+    if (Math.abs(start) > len || Math.abs(end) > len || startIdx > endIdx) {
+      return Nothing();
+    }
+    return Just(R.slice(startIdx, endIdx, xs));
```
Comment (davidchambers) [inline index.js]: I like to avoid using `return` for control flow where possible. Let's write something like this instead:

```javascript
return (Math.abs(start) <= len &&
        Math.abs(end) <= len &&
        startIdx <= endIdx) ?
  Just(R.slice(startIdx, endIdx, xs)) :
  Nothing();
```

I'm sorry about the nitpicking!

File: test/index.js [hunk 8]
```diff
@@ -1117,6 +1118,60 @@ describe('list', function() {
 
   });
 
+  describe('slice', function() {
+
+    it('returns a Nothing if applied to an empty list', function() {
+      assert(S.slice(0, 3, []).equals(S.Nothing()));
+    });
```
Comment (davidchambers) [inline test/index.js]: This is misleading since `S.slice(0, 0, [])` evaluates to `Just([])`. I think we could remove this assertion.

File: index.js (hunk 7 above)
Comment (svozza) [inline index.js]: No bother. Is there a particular reason to avoid that pattern? Genuinely interested to know because I do it a lot.

File: test/index.js [hunk 9]
```diff
@@ -1117,6 +1118,60 @@ describe('list', function() {
...
+    it('returns a Nothing with a negative end index greater than list length', function() {
+        assert(S.slice(1, -6, [1, 2, 3, 4, 5]).equals(S.Nothing()));
+      });
+
+    it('returns a Nothing with a negative start index greater than list length', function() {
+      assert(S.slice(-6, 1, [1, 2, 3, 4, 5]).equals(S.Nothing()));
+    });
+
+    it('returns a Just with an empty array when start index equals end index', function() {
+        eq(S.slice(1, 1, [1, 2, 3, 4, 5]).toString(), 'Just([])');
```
Comment (davidchambers) [inline test/index.js]: This could be expanded to eight assertions:

```javascript
eq(S.slice(1, 1, [1, 2, 3, 4, 5]).toString(), 'Just([])');
eq(S.slice(1, -4, [1, 2, 3, 4, 5]).toString(), 'Just([])');
eq(S.slice(-4, 1, [1, 2, 3, 4, 5]).toString(), 'Just([])');
eq(S.slice(-4, -4, [1, 2, 3, 4, 5]).toString(), 'Just([])');
eq(S.slice(0, 0, []).toString(), 'Just([])');
eq(S.slice(0, -0, []).toString(), 'Just([])');
eq(S.slice(-0, 0, []).toString(), 'Just([])');
eq(S.slice(-0, -0, []).toString(), 'Just([])');
```

File: test/index.js [hunk 10]
```diff
@@ -1117,6 +1118,60 @@ describe('list', function() {
...
 
+  describe('slice', function() {
+
+    it('returns a Nothing if applied to an empty list', function() {
+      assert(S.slice(0, 3, []).equals(S.Nothing()));
+    });
+
+    it('returns a Nothing with a positive end index greater than start index', function() {
+        assert(S.slice(6, 1, [1, 2, 3, 4, 5]).equals(S.Nothing()));
+      });
```
Comment (davidchambers) [inline test/index.js]: :triangular_ruler: The indentation is a bit off here.

Comment (svozza) [inline test/index.js]: Fixed.

File: test/index.js (hunk 9 above)
Comment (svozza) [inline test/index.js]: Good call!

File: index.js (hunk 7 above)
Comment (davidchambers) [inline index.js]: Returning early isn't bad or wrong, it's simply an imperative construct. My preference is to avoid imperative constructs, though in some cases they are very useful.

In this particular case we're not returning early to avoid unnecessary work, we're simply *branching*. I believe the indentation should match the logical branching, so I advocate

```javascript
if (x) {
  return y;
} else {
  return z;
}
```

rather than

```javascript
if (x) {
  return y;
}
return z;
```

`if`/`else` is another imperative construct: it's not an expression, so it's only useful for side effects. So I like `return x ? y : z;` best of all.

File: test/index.js [hunk 11]
```diff
@@ -1117,6 +1118,63 @@ describe('list', function() {
...
+    });
+
+    it('returns a Just with a negative end index', function() {
+      eq(S.slice(1, -2, [1, 2, 3, 4, 5]).toString(), 'Just([2, 3])');
+    });
+
+    it('accepts -0 for a start index as the position half a step beyond the last index',
+      function() {
+        eq(S.slice(2, -0, [1, 2, 3, 4, 5]).toString(), 'Just([3, 4, 5])');
+    });
```
Comment (davidchambers) [inline test/index.js]: Should the description say _end index_ rather than _start index_?

File: index.js (hunk 7 above)
Comment (svozza) [inline index.js]: Interesting, I've not thought of if/else in those terms even though it's blindingly obviously when you explain like you have there.

File: test/index.js (hunk 11 above)
Comment (svozza) [inline test/index.js]: It should indeed, I mixed the body of the two tests up there. I've actually rolled them into one test block now.

File: index.js (hunk 7 above)
Comment (davidchambers) [inline index.js]: It was only recently that I read or heard it put this way. It was an aha moment for me too. :)

File: test/index.js [hunk 12]
```diff
@@ -1117,6 +1118,59 @@ describe('list', function() {
...
+    it('returns a Just with a negative start index', function() {
+      eq(S.slice(-3, 5, [1, 2, 3, 4, 5]).toString(), 'Just([3, 4, 5])');
+    });
+
+    it('returns a Just with a negative end index', function() {
+      eq(S.slice(1, -2, [1, 2, 3, 4, 5]).toString(), 'Just([2, 3])');
+    });
+
+    it('accepts -0 as the position half a step beyond the last index',
+      function() {
```
Comment (davidchambers) [inline test/index.js]: Let's remove this line wrapping.

Comment (svozza) [inline test/index.js]: Yeah, I didn't like it either, I only had it that way because it was previously over 79 chars.

**Comment by davidchambers** [comment]
:deciduous_tree: Thanks for persevering, @svozza. :)

**Comment by svozza** [comment]
Well, there were 47 comments on my last PR so I'm taking this as an improvement! My next goal is to get it under 20. :)

File: index.js (hunk 1 above)
Comment (svozza) [inline index.js]: Okay, so for a laugh I perserved with this. :smiley: 

```javascript
  var at = S.at = R.curry(R.compose(R.map(R.head), R.useWith(R.uncurryN(2, R.converge(S.slice, R.identity, R.ifElse(R.eq(-1), R.always(-0), R.inc))), R.identity, R.identity)));
```

Comment (svozza) [inline index.js]: This might be somewhat easier to read(!):

```javascript
var at = S.at = R.curry(R.pipe(
    R.useWith(
      R.uncurryN(2, R.converge(
        S.slice,
        R.identity,
        R.ifElse(R.eq(-1), R.always(-0), R.inc))
      ),
      R.identity,
      R.identity),
    R.map(R.head)))
```

Comment (davidchambers) [inline index.js]: That's quite something! Here it is reformatted:

```javascript
var at = S.at =
R.curry(R.compose(R.map(R.head),
                  R.useWith(R.uncurryN(2,
                                       R.converge(S.slice,
                                                  R.identity,
                                                  R.ifElse(R.eq(-1),
                                                           R.always(-0),
                                                           R.inc))),
                            R.identity,
                            R.identity)));
```

Sometimes a lambda is just the ticket. :)

Comment (svozza) [inline index.js]: I think what this really proves is that I have _far_ too much time on my hands.

Comment (svozza) [inline index.js]: Actually, funnily enough, the thing I wasted most time on (aside from the whole enterprise, that is) was trying to get all the tests running only to finally realise that it was because `compose` is not autocurried.
